// @openapi-codegen/typescript@3.0.1/lib/templates/fetcher.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getFetcher = void 0;
const case_1 = require("case");
/**
 * Get fetcher template
 *
 * @param contextPath import the context from another file
 */
const getFetcher = ({ prefix, contextPath, baseUrl, }) => `${contextPath
    ? `import { ${(0, case_1.pascal)(prefix)}Context } from "./${contextPath}";`
    : `export type ${(0, case_1.pascal)(prefix)}FetcherExtraProps = {
      /**
       * You can add some extra props to your generated fetchers.
       * 
       * Note: You need to re-gen after adding the first property to
       * have the \`${(0, case_1.pascal)(prefix)}FetcherExtraProps\` injected in \`${(0, case_1.pascal)(prefix)}Components.ts\`
       **/
    }`}

const baseUrl = ${baseUrl ? `"${baseUrl}"` : `""; // TODO add your baseUrl`}

export type ${(0, case_1.pascal)(prefix)}FetcherOptions<TBody, THeaders, TQueryParams, TPathParams> = {
  url: string;
  method: string;
  body?: TBody;
  headers?: THeaders;
  queryParams?: TQueryParams;
  pathParams?: TPathParams;
} & ${contextPath
    ? `${(0, case_1.pascal)(prefix)}Context["fetcherOptions"];`
    : `${(0, case_1.pascal)(prefix)}FetcherExtraProps`}

export async function ${(0, case_1.camel)(prefix)}Fetch<
  TData,
  TBody extends {} | undefined | null,
  THeaders extends {},
  TQueryParams extends {},
  TPathParams extends {}
>({
  url,
  method,
  body,
  headers,
  pathParams,
  queryParams,
}: ${(0, case_1.pascal)(prefix)}FetcherOptions<
  TBody,
  THeaders,
  TQueryParams,
  TPathParams
>): Promise<TData> {
  const response = await window.fetch(\`\${baseUrl}\${resolveUrl(url, queryParams, pathParams)}\`,
    {
      method: method.toUpperCase(),
      body: body ? JSON.stringify(body) : undefined,
      headers: {
        "Content-Type": "application/json",
        ...headers,
      },
    }
  );
  if (!response.ok) {
    // TODO validate & parse the error to fit the generated error types 
    throw new Error("Network response was not ok");
  }
  return await response.json();
}

const resolveUrl = (
  url: string,
  queryParams: Record<string, string> = {},
  pathParams: Record<string, string> = {}
) => {
  let query = new URLSearchParams(queryParams).toString();
  if (query) query = \`?\${query}\`;
  return url.replace(/\\{\\w*\\}/g, (key) => pathParams[key.slice(1, -1)]) + query;
};
`;
exports.getFetcher = getFetcher;
